Inductor Freezing

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100746
* #100653
* __->__ #100652


Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov @soumith @desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

eellison committed:
Inductor Optimize For Inference/Freezing

[ghstack-poisoned]

eellison committed:
Update on "Inductor Optimize For Inference/Freezing"




cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Optimize For Inference/Freezing"




cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Optimize For Inference/Freezing"




cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Optimize For Inference/Freezing"




cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Optimize For Inference/Freezing"


Gives 1% boost on hf_Bert inference.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Optimize For Inference/Freezing"


Gives 1% boost on hf_Bert inference.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"


Gives 1% boost on hf_Bert inference.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor. This occurs post functionalization in aot autograd. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used an eager. 
- Currently there is no way in aot autograd to preserve the original calling convention, which prevents discarding the previous memory (cc bdhirsh)
- Caching parameter transformations/constant folding across different inferences nyi 


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor. This occurs post functionalization in aot autograd. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used an eager. 
- Currently there is no way in aot autograd to preserve the original calling convention, which prevents discarding the previous memory (cc bdhirsh)
- Caching parameter transformations/constant folding across different inferences nyi 


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor. This occurs post functionalization in aot autograd. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used an eager. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be.. Could be freezing or something else.


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor. This occurs post functionalization in aot autograd. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used an eager. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be.. Could be freezing or something else.


cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46244033).

leslie-fang-intel commented:
Hi @eellison, Thanks for providing these passes. There is a use case in Quantization 2.0 with Inductor backend: https://github.com/pytorch/pytorch/blob/040d2cc96953290bfe49e32dbb250fa9df6f51ce/test/quantization/pt2e/test_quantize_pt2e_fx.py#L325-L446

It will use: 
1. `torchdynamo.export` to capture the fx graph.
2. Pass the quantization 2.0 flow to generate reference quantized module. 
3. Use `compile_fx` lower into Inductor with the reference quantized module. 

As discussed here https://github.com/pytorch/pytorch/pull/101164#discussion_r1195953417, we depend on the `constant folding` pass to do further optimization. So I take a try with your PR. However, I got error message as
```
  File "/home/lesliefang/pytorch_1_7_1/inductor_quant/pytorch/torch/_functorch/aot_autograd.py", line 2273, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/home/lesliefang/pytorch_1_7_1/inductor_quant/pytorch/torch/_functorch/aot_autograd.py", line 1535, in aot_dispatch_base
    compiled_fw = compiler(fw_module, flat_args)
  File "/home/lesliefang/pytorch_1_7_1/inductor_quant/pytorch/torch/_inductor/compile_fx.py", line 710, in inference_compiler
    fw_metadata=torch._guards.TracingContext.get().fw_metadata,
AttributeError: 'NoneType' object has no attribute 'fw_metadata'
```

Could you take a look of this issue? I create a [gist here](https://gist.github.com/leslie-fang-intel/c1e1fe49ae9478425d39e14775fde7ea) to reproduce it.

bdhirsh commented on torch/_inductor/compile_fx.py:
```diff
@@ -693,7 +694,55 @@ def fw_compiler_base(model: torch.fx.GraphModule, example_inputs, is_inference):
+            aot_autograd_model: torch.fx.GraphModule, example_inputs
+        ):
+            # partition_fn won't be called
+            joint_graph_passes(aot_autograd_model)
```
It looks like before this PR, none of the joint_graph_passes would run when inductor sees an inference graph. Do we need to run them now for inference, but only when using `config.freezing`? (and more of a curiosity question - why?)

bdhirsh replied:
oh i see, `fw_compiler_base()` calls the joint passes during inference. I guess you can't easily re-use it, since you wnat to run the joint passes before doing the freezing, but call `inner_compile` after the freezing.

And (answering my own question) it looks like we're guaranteed that the joint passes will always run- it looks like `fw_compiler` here doesn't invoke the joint passes, but we're guaranteed in aot autograd that we'll be running the partition_fn in that case, which also runs the passees.

bdhirsh commented on torch/_inductor/compile_fx.py:
```diff
@@ -693,7 +694,55 @@ def fw_compiler_base(model: torch.fx.GraphModule, example_inputs, is_inference):
+                )
+
+            # Need to drop the args we have constant-ified.
+            params_flat = torch._guards.TracingContext.get().params_flat
```
Maybe this should happen inside of `freeze()` too, so all of the calling convention changes happen in one place?

bdhirsh commented on torch/_inductor/freezing.py:
```diff
@@ -0,0 +1,208 @@
+        of the inputs that were preserved (not turned into constants).
+    """
+    params_flat = torch._guards.TracingContext.get().params_flat
+    preserved_arg_indices = replace_params_with_constants(
```
another curiosity question - do you imagine that shape padding of parameters would also (maybe) happen here?

eellison replied:
It happens naturally as part of the constant folding

jansel commented on torch/_inductor/freezing.py:
```diff
@@ -0,0 +1,208 @@
+
+    node_replacements = {}
+
+    class ConstantFolder(torch.fx.Interpreter):
```
Will this work for dynamic shapes?

I am thinking something like `torch.zeros([SymInt, SymInt])` might get constant folded?

Also, thoughts on if we should constant fold `zeros`, `ones`, `full`, etc?
1) If they are used by other constant ops -- especially fallback ones or params, this seems like a good folding to do
2) If they are used by non-constant ops -- inductor will just turn them into a constant in the generated kernel so folding will result in an extra load.

eellison replied:
Good points. I'll double check that Symbolic shapes works. 

For `zeros`, `ones`, `full` etc - we could check uses, we could also add logic so that constants with a uniform value get mapped to `Constant`. I think the latter would be simpler. We can pass along that certain constants are uniform in the meta.

jansel replied:
There is also `torch.arange` which inductor can codegen without any load.

I think we likely don't want to constant fold these, because I suspect inductor generates near optimal code for them.   Though there certainly are exceptions.  Might be worth benchmarking.

eellison replied:
I just skipped constructors for now, will come back to this and benchmark different approaches

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/freezing.py:
```diff
@@ -0,0 +1,208 @@
+
+            # TODO - remove constant from node_replacement when it has no uses
+            if node.op != "get_attr" and isinstance(out, torch.Tensor):
+                node_replacements[node] = out
```
@eellison Do you think is it reasonable adding some check here to filter nodes we don't want to do the constant folding? For example, take quantization conv as example, there is a pattern `fp32_weight->quant_per_channel->dequant_per_channel->conv`. Although `fp32_weight->quant_per_channel->dequant_per_channel` can do constant folding into new fp32 weight, however, we want to keep `dequant_per_channel` node in the graph. Then `fp32_weight->quant_per_channel` can be constant folding into a int8_weight, and `dequant_per_channel->conv` can be further lowering into a `int8_conv` node. Also cc @jgong5 @Xia-Weiwen.

eellison replied:
Yea, makes sense, we can deal with this and the other quantization issue in a follow up

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov soumith desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov soumith desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison committed:
Update on "Inductor Freezing"



Adds a freezing pass that will constant fold parameters in inductor `config.freezing`. This occurs post functionalization in aot autograd to capture both dispatching and allow passes to occur post functionalization. A few notes:

- There is an option to discard parameters `config.freezing_discard_parameters` which will take the current eager modules and wrap parameters to a Tensor subclass which will error if used. 
- I needed to expose flat_params in aot_autograd in order to discard old references when we constant fold away parameters, like with amp. I also exposed `fw_metadata` to avoid constant folding mutated paraemters. 
- Caching parameter transformations/constant folding across different inferences nyi
- Checking version_counter of constant folded params nyi


I'm not really sure what the actual naming should be. In jit there was both "freezing", which was platform agnostic, and "optimize for inference", which made device specific optimizations. We're doing the latter here but maybe freezing is a better name.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov soumith desertfire

Differential Revision: [D46244033](https://our.internmc.facebook.com/intern/diff/D46244033)

[ghstack-poisoned]

eellison commented:
Test disabled here https://github.com/pytorch/pytorch/issues/103466

huydhn commented:
There are now other failed tests besides the above disabled one https://hud.pytorch.org/pytorch/pytorch/commit/c3d3165f16dccd88872139b72cd421e0ceafdd9b, and the diff hasn't been landed internally yet, so should we submit a revert PR or disable the whole test_inductor_freezing file?

eellison commented:
@huydhn let's disable the file if that works for you. it's not turned on or used anywhere yet
